trantor/net/inner/tlsprovider/BotanTLSProvider.cc

```diff
@@ -0,0 +1,390 @@
+                    Botan::TLS::Alert::BAD_CERTIFICATE,
+                    "Certificate does not match hostname");
+            }
+            if (policyPtr_->getValidateDate())
+            {
+                auto notBefore = cert.not_before();
+                auto notAfter = cert.not_after();
+                auto now = Botan::ASN1_Time(std::chrono::system_clock::now());
+                if (now < notBefore || now > notAfter)
+                {
+                    handleSSLError(SSLError::kSSLInvalidCertificate);
+                    throw std::runtime_error(
+                        "Certificate is not valid for current time");
+                }
+            }
+            setPeerCertificate(std::make_shared<BotanCertificate>(cert));
+        }
+
+        tlsConnected_ = true;
+        loop_->queueInLoop([this]() {
+            setApplicationProtocol(channel_->application_protocol());
+            if (handshakeCallback_)
+                handshakeCallback_(conn_);
+        });
+        // Do we want to cache all sessions?
+        return true;
+    }
+
+    void tls_verify_cert_chain(
+        const std::vector<Botan::X509_Certificate> &certs,
+        const std::vector<std::shared_ptr<const Botan::OCSP::Response>> &ocsp,
+        const std::vector<Botan::Certificate_Store *> &trusted_roots,
+        Botan::Usage_Type usage,
+        const std::string &hostname,
+        const Botan::TLS::Policy &policy)
+    {
+        setSniName(hostname);
+        if (policyPtr_->getValidateChain())
+            Botan::TLS::Callbacks::tls_verify_cert_chain(
+                certs, ocsp, trusted_roots, usage, hostname, policy);
+    }
+
+    TrantorPolicy validationPolicy_;
+    std::unique_ptr<Botan::Credentials_Manager> credsPtr_;
+    std::unique_ptr<Botan::TLS::Channel> channel_;
+    bool tlsConnected_ = false;
+    ssize_t lastWriteSize_ = 0;
+};
+
+std::unique_ptr<TLSProvider> trantor::newTLSProvider(TcpConnection *conn,
+                                                     TLSPolicyPtr policy,
+                                                     SSLContextPtr ctx)
+{
+    return std::make_unique<BotanTLSProvider>(conn,
+                                              std::move(policy),
+                                              std::move(ctx));
+}
+
+SSLContextPtr trantor::newSSLContext(const TLSPolicy &policy, bool server)
+{
+    auto ctx = std::make_shared<SSLContext>();
+    ctx->isServer = server;
+    if (!policy.getKeyPath().empty())
+    {
+        Botan::DataSource_Stream in(policy.getKeyPath());
+        ctx->key = Botan::PKCS8::load_key(in);
+    }
+
+    if (!policy.getCertPath().empty())
+    {
+        ctx->cert =
+            std::make_unique<Botan::X509_Certificate>(policy.getCertPath());
+    }
+
+    if (!policy.getCaPath().empty())
+    {
+        ctx->certStore = std::make_unique<Botan::Flatfile_Certificate_Store>(
+            policy.getCaPath());
+        if (!server)
+            ctx->requireClientCert = true;
```

Comment:
Just a wild guess, but shouldn't that require client authentication if it is a server? Rather than if it is not?